### DISCUSSION TITLE
关于原Shoka评论数据转移至Waline是否可行

### DISCUSSION THREAD

**Discussion by MqyGalaxy** [theme-shoka-x/hexo-theme-shokaX#311] [Q&A]
在配置Waline时，我直接使用了之前在Shoka的MiniValine所使用的LeadCloud数据库。原本Waline应该可以直接使用Valine的数据，但貌似Shoka作者魔改后的MiniValine的评论数据在 `hexo s` 后无法被Waline正常读取：

![image](https://github.com/user-attachments/assets/4c847536-90e3-4a1d-88cb-d4ef74a17f09)

如图，第一个评论是在部署完Waline后直接使用Waline发布的评论，可以正常在博客中显示以及跳转。而第二个评论是由原Shoka的MiniValine发布的评论，无法在文章页显示，以及在最新评论中显示错误，无法正确跳转。

在服务端以及LeanCloud数据库中发现以下问题：

![屏幕截图 2024-07-25 060639](https://github.com/user-attachments/assets/3293f694-c669-4d8a-9af5-ce267ad8e549)

 - 在Waline的服务端中看到url是错误的，目前已经明白由于 `url` 的格式不同导致原本在MiniValine的评论无法在Waline中显示（路径没有定位到根目录）。

 - 在LeanCloud后台手动将 `url` 修改过后就正常在文章中显示了，但由于Shoka的 `comment` 直接写入了html格式，在“最新评论”中显示的效果也很奇怪。

 - 文章阅读数的数据因为 ~xid~ `url` 同样没有定位到根目录的原因也无法正常读取。

 - 根据[这位大佬的文章所说](https://linn-ylz.com/Hexo/transfer-waline/#%E8%AF%84%E8%AE%BA)因为Waline是通过 `insertedAt` 列来读取评论时间的，而Shoka的MiniValine是从 `createdAt` 列读取时间，并没有在 `insertedAt` 中写入时间。这导致 `insertedAt` 为空从而无法使Waline读取到评论时间。

想问一下大家有没有把Shoka的评论数据迁移到ShokaX的具体案例，由于本人的评论数据还挺多的，不是很舍得放弃，手动调整上百条评论和文章阅读数感觉也不是一件很现实的事情。

**Comment by zkz098** [comment]
我有一个不太靠谱的想法
![image](https://github.com/user-attachments/assets/941c91c0-0a59-406a-9040-b0c1bdfadd2b)
先将valine数据转成twikoo/artalk，然后再把转换后数据转换回waline有没有可能解决，因为valine 和 waline LeanCloud不能互转，走一个第三方重新处理一下数据库（）

附链接：
1. https://artalk.js.org/guide/transfer.html
2. https://waline.js.org/migration/tool.html

**Reply by MqyGalaxy** [reply]
貌似不行，转换成artalk再转换成waline会报错💦
![image](https://github.com/user-attachments/assets/a79e6d6a-2e7a-4b7f-ae28-a4ea86836bd4)
而且路径格式貌似没有变化，感觉就算转换完成也不会起作用。

**Reply by zkz098** [reply]
有时间可以把评论数据发送到 `zkz20081204#126.com`(#改@)，我看看能不能手写一个迁移脚本，从Shoka Minivaline 4.x 到 Waline 3.x等于跨了3-5个大版本，兼容可能性无限趋近于0（

**Reply by MqyGalaxy** [reply]
主要是数据库格式不同，如果能通过脚本修改 `url` 链接格式和添加 `insertedAt` 应该就能解决问题。

Shoka 主题的 MiniValine 使用html格式倒是比较麻烦，会导致在最近评论中出现直接将老评论的h5格式显示出来的问题，以及老评论回复会多一个 `@受到回复的人` ，因为 Shoka MiniValine 也把这个写在 `comment` 评论内容的h5里了。但主要还是先将数据转移了比较好。
